### PR TITLE
Replace json serialization with faster string-stream based implementation

### PR DESCRIPTION
**Pull request by johannes-wolf**
### Changes

Replace json serialization with a custom implementation – speedyj – that does not build a DOM, but serializes directly into a string stream.

### Review Checklist

- [ ] The changes are understood.
- [ ] The changes are well documented.
- [ ] The changes have been tested.

### CODE REVIEW

File: runtime/zsr/introspectable-json.hpp [hunk 1]
```diff
@@ -47,8 +47,8 @@ enum ZSR_EXPORT SerializationOptions
  *       }
  *     }
  */
-nlohmann::json ZSR_EXPORT serialize(const zsr::Introspectable& obj,
-                                    const std::deque<zsr::Package>* pkgs = nullptr,
-                                    SerializationOptions opts = SERIALIZE_DEFAULT);
+speedyj::Stream ZSR_EXPORT serialize(const zsr::Introspectable& obj,
```
Comment (josephbirkner) [inline runtime/zsr/introspectable-json.hpp:61]: The docstring should must be updated I think 📜

File: runtime/zsr/speedy-j.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Internal stream state.
+ */
+struct StreamState
+{
+    enum Type {
+        Object,
+        Array
+    } type;
+    int itemIdx;
```
Comment (josephbirkner) [inline runtime/zsr/speedy-j.hpp]: Please initialize inline 😺

File: runtime/src/speedy-j.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+namespace speedyj
+{
+
+static std::string jsonEncoded(const std::string& s)
+{
+    std::string res;
+    res.reserve(s.size());
+
+    for (auto c : s) {
```
Comment (josephbirkner) [inline runtime/src/speedy-j.cpp]: Small nitpick: `auto const&`? Or do you think this should definitely be caught by the optimizer?

**Review by josephbirkner — APPROVED** [review]
```
                                         _________         ________ 
            ________________ _____ _____ ______  /_____  ________(_)
            __  ___/___  __ \_  _ \_  _ \_  __  / __  / / /_____  / 
            _(__  ) __  /_/ //  __//  __// /_/ /  _  /_/ / ____  /  
            /____/  _  .___/ \___/ \___/ \__,_/   _\__, /  ___  /   
                    /_/                           /____/   /___/    
            
```

File: runtime/src/speedy-j.cpp (hunk 3 above)
Comment (johannes-wolf) [reply runtime/src/speedy-j.cpp]: Variable `c` is of type `char`. I can't think of any optimization the optimizer could apply. As far as I know `auto` must _never_ deduce to a reference type.

File: runtime/src/speedy-j.cpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        case '"':
+            res += "\\\"";
+            break;
+
+        case '\\':
+            res += "\\\\";
+            break;
+
+        default:
+            if (c <= ' ') {
```
Comment (johannes-wolf) [inline runtime/src/speedy-j.cpp]: Change to `<`; escaping spaces blows up document size.

File: runtime/src/speedy-j.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+namespace speedyj
+{
+
+static std::string jsonEncoded(const std::string& s)
+{
+    std::string res;
+    res.reserve(s.size());
+
+    for (auto c : s) {
+        switch (c) {
```
Comment (johannes-wolf) [inline runtime/src/speedy-j.cpp:14]: Add special escapes for: `\n \r \f \b` and `\t`.
See https://www.ietf.org/rfc/rfc4627.txt page 4.

File: runtime/src/speedy-j.cpp [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    }
+}
+
+static void finish(Stream& s)
+{
+    switch (s.state().type) {
+    case StreamState::Array:
+        s.ss_ << ']';
+        break;
+    case StreamState::Object:
```
Comment (johannes-wolf) [inline runtime/src/speedy-j.cpp:104]: Assert that `itemIdx` is 0 or even (otherwise an invalid object has been serialized).

**Review by josephbirkner — APPROVED** [review]
LGTM

**Pull request merged**